fdmi/plugins/m0scheduler

```diff
@@ -17,128 +17,277 @@
 # For any questions about this software or licensing,
 # please email opensource@seagate.com or cortx-questions@seagate.com.
 #
+
+# This is FDMI plugin sample app wrapper that serves as a scheduler
+# for the FDMI records. It gathers that cluster infp using Hare hctl
+# tool, forks m0sched program and waits for records delivered by it.
+# For now it is just a demo that prints FDMI records to console, but
+# later it can perform records de-dup and replication coordination by
+# means of set of the replication python programs that perform actual
+# data move.
+
 import subprocess
 import signal
-from socket import gethostname
 import shlex
 import sys
 import json
-import ast
 import codecs
 
-kv_records = [] #global kv cache
+cluster_info = dict()  # global cluster info
+kv_records = []        # global kv cache
+processes = []         # global list if m0sched processes
+
 
 def str_decode(encoded):
-	if encoded == "0" :
-		return encoded;
-	#return codecs.decode(encoded, 'hex').decode('utf-8')
-	return codecs.decode(encoded, 'hex')
+    if encoded is None:
+        return None
+    if encoded == '0':
+        return encoded
+    # return codecs.decode(encoded, 'hex').decode('utf-8')
+    return codecs.decode(encoded, 'hex')
+
 
 def json_parser(kv_record):
-	kvs = json.loads(kv_record)
-	kv = {'fid': str_decode(kvs["fid"]), 'cr_key': str_decode(kvs["cr_key"]), 'cr_val': str_decode(kvs["cr_val"])}
-	#kv = {'fid': kvs["fid"], 'cr_key': kvs["cr_key"], 'cr_val': kvs["cr_val"]}
-	flag =  True
-
-	#check dedup records
-	for kv_i in kv_records:
-		if kv['cr_key'] == kv_i['cr_key'] and kv['cr_val'] != "0":
-			#print(kv_records)
-			print("dedup: ", kv_i, kv)
-			flag = False
-			break
-
-	#if  record is not dedupable, add to cache
-	if flag == True and kv['cr_val'] != "0":
-		print("new kv: ", kv)
-		kv_records.append(kv)
-
-#handle SIGNT and SIGTERM
+    kvs = json.loads(kv_record)
+    fid = str_decode(kvs.get('fid'))
+    if fid is None:
+        print('Invalid fid in kv_record {}'.format(kv_record), file=sys.stderr)
+        return
+    cr_key = str_decode(kvs.get('cr_key'))
+    if cr_key is None:
+        print('Invalid cr_key in kv_record {}'.format(kv_record), file=sys.stderr)
+        return
+    cr_val = str_decode(kvs.get('cr_val'))
+    if cr_val is None:
+        print('Invalid cr_val in kv_record {}'.format(kv_record), file=sys.stderr)
+        return
+    kv = {
+        'fid': fid,
+        'cr_key': cr_key,
+        'cr_val': cr_val
+    }
+    # kv = {'fid': kvs["fid"], 'cr_key': kvs["cr_key"], 'cr_val': kvs["cr_val"]}
+    flag = True
+
+    # check dedup records
+    for kv_i in kv_records:
```

Comment:
We can change kv_records to be dictionary with key as `kv['cr_key']` so this can change to dictionary lookup.

Reply:
agreed